### PR TITLE
RFR: API to cancel an ongoing execution using DELETE

### PR DESCRIPTION
**Pull request by lakshmi-kannan**
## API:

```DELETE http://localhost:9101/v1/actionexecutions/5528162e6d0aff79e00d5296```

## TODO:
[ ] - CLI exposure for cancel execution. I prefer st2 execution {id} --cancel. We should explicitly state a message in CLI that it's a best effort to cancel. 

## Sample:

```
~ ❯❯❯ st2 execution list
+----------------------------+------------------+--------------+---------+------------------+---------------+
| id                         | action.ref       | context.user | status  | start_timestamp  | end_timestamp |
+----------------------------+------------------+--------------+---------+------------------+---------------+
| + 5528162e6d0aff79e00d5296 | examples.echocha | stanley      | running | Fri, 10 Apr 2015 |               |
|                            | in               |              |         | 18:27:58 UTC     |               |
+----------------------------+------------------+--------------+---------+------------------+---------------+
~ ❯❯❯ st2 execution get 5528162e6d0aff79e00d5296 --tasks
id: 5528162e6d0aff79e00d5296
action.ref: examples.echochain
status: running
start_timestamp: 2015-04-10T18:27:58.575250Z
end_timestamp: None
+--------------------------+-----------+------+------------+----------------------+
| id                       | status    | task | action     | start_timestamp      |
+--------------------------+-----------+------+------------+----------------------+
| 5528162e6d0aff79e5103104 | succeeded | c1   | core.local | Fri, 10 Apr 2015     |
|                          |           |      |            | 18:27:58 UTC         |
| 5528162f6d0aff79e5103107 | running   | c2   | core.local | Fri, 10 Apr 2015     |
|                          |           |      |            | 18:27:59 UTC         |
+--------------------------+-----------+------+------------+----------------------+
~ ❯❯❯ st2 execution get 5528162e6d0aff79e00d5296 --tasks
~ ❯❯❯ http DELETE http://localhost:9101/v1/actionexecutions/5528162e6d0aff79e00d5296                                        ⏎
HTTP/1.1 200 OK
Access-Control-Allow-Headers: Content-Type,Authorization,X-Auth-Token
Access-Control-Allow-Methods: GET,POST,PUT,DELETE,OPTIONS
Access-Control-Allow-Origin: http://localhost:3000
Access-Control-Expose-Headers: Content-Type,X-Limit,X-Total-Count
Connection: keep-alive
Content-Length: 989
Content-Type: application/json; charset=UTF-8
Date: Fri, 10 Apr 2015 18:28:26 GMT

{
    "action": {
        "description": "Simple Action Chain workflow",
        "enabled": true,
        "entry_point": "chains/echochain.yaml",
        "id": "5528162a6d0aff79fe81fb7d",
        "name": "echochain",
        "pack": "examples",
        "parameters": {},
        "ref": "examples.echochain",
        "runner_type": "action-chain",
        "tags": []
    },
    "children": [
        "5528162e6d0aff79e5103104",
        "5528162f6d0aff79e5103107"
    ],
    "context": {
        "user": "stanley"
    },
    "end_timestamp": "2015-04-10T18:28:26.436761Z",
    "id": "5528162e6d0aff79e00d5296",
    "liveaction": {
        "action": "examples.echochain",
        "callback": {},
        "id": "5528162e6d0aff79e00d5295",
        "runner_info": {
            "hostname": "vagrant-ubuntu-utopic-64",
            "pid": 31205
        }
    },
    "result": {
        "message": "Action canceled by user."
    },
    "runner": {
        "description": "A runner for launching linear action chains.",
        "enabled": true,
        "id": "552816286d0aff79fe81fab1",
        "name": "action-chain",
        "runner_module": "st2actions.runners.actionchainrunner",
        "runner_parameters": {}
    },
    "start_timestamp": "2015-04-10T18:27:58.575250Z",
    "status": "canceled"
}

~ ❯❯❯ st2 execution get 5528162e6d0aff79e00d5296 --tasks
id: 5528162e6d0aff79e00d5296
action.ref: examples.echochain
status: canceled
start_timestamp: 2015-04-10T18:27:58.575250Z
end_timestamp: 2015-04-10T18:28:26.436761Z
+--------------------------+-----------+------+------------+----------------------+
| id                       | status    | task | action     | start_timestamp      |
+--------------------------+-----------+------+------------+----------------------+
| 5528162e6d0aff79e5103104 | succeeded | c1   | core.local | Fri, 10 Apr 2015     |
|                          |           |      |            | 18:27:58 UTC         |
| 5528162f6d0aff79e5103107 | running   | c2   | core.local | Fri, 10 Apr 2015     |
|                          |           |      |            | 18:27:59 UTC         |
+--------------------------+-----------+------+------------+----------------------+
~ ❯❯❯ st2 execution get 5528162e6d0aff79e00d5296 --tasks
id: 5528162e6d0aff79e00d5296
action.ref: examples.echochain
status: canceled
start_timestamp: 2015-04-10T18:27:58.575250Z
end_timestamp: 2015-04-10T18:28:26.436761Z
+--------------------------+-----------+------+------------+----------------------+
| id                       | status    | task | action     | start_timestamp      |
+--------------------------+-----------+------+------------+----------------------+
| 5528162e6d0aff79e5103104 | succeeded | c1   | core.local | Fri, 10 Apr 2015     |
|                          |           |      |            | 18:27:58 UTC         |
| 5528162f6d0aff79e5103107 | succeeded | c2   | core.local | Fri, 10 Apr 2015     |
|                          |           |      |            | 18:27:59 UTC         |
+--------------------------+-----------+------+------------+----------------------+
~ ❯❯❯
```

```
(virtualenv)~/st2 git:STORM-999/stop_execution ❯❯❯ st2 run examples.echochain                                         ⏎ ✭ ✱ ◼
..............
id: 5528162e6d0aff79e00d5296
action.ref: examples.echochain
status: canceled
start_timestamp: 2015-04-10T18:27:58.575250Z
end_timestamp: 2015-04-10T18:28:26.436761Z
+--------------------------+-----------+------+------------+----------------------+
| id                       | status    | task | action     | start_timestamp      |
+--------------------------+-----------+------+------------+----------------------+
| 5528162e6d0aff79e5103104 | succeeded | c1   | core.local | Fri, 10 Apr 2015     |
|                          |           |      |            | 18:27:58 UTC         |
| 5528162f6d0aff79e5103107 | running   | c2   | core.local | Fri, 10 Apr 2015     |
|                          |           |      |            | 18:27:59 UTC         |
+--------------------------+-----------+------+------------+----------------------+
(virtualenv)~/st2 git:STORM-999/stop_execution ❯❯❯
```

### CODE REVIEW

File: st2actions/st2actions/worker.py [hunk 1]
```diff
@@ -74,6 +75,8 @@ def _do_process_task(self, body):
 
     def execute_action(self, liveaction):
         # Note: We only want to execute actions which haven't completed yet
+        if liveaction.status == LIVEACTION_STATUS_CANCELED:
+            LOG.info('Not executing liveaction %s. User canceled execution.', liveaction.id)
```
Comment (manasdk) [inline st2actions/st2actions/worker.py]: missing a return?

File: st2common/st2common/constants/action.py [hunk 2]
```diff
@@ -40,9 +41,13 @@
...
 
 LIVEACTION_STATUSES = [LIVEACTION_STATUS_SCHEDULED,
                        LIVEACTION_STATUS_RUNNING, LIVEACTION_STATUS_SUCCEEDED,
-                       LIVEACTION_STATUS_FAILED]
+                       LIVEACTION_STATUS_FAILED, LIVEACTION_STATUS_CANCELED]
 
 ACTION_OUTPUT_RESULT_DELIMITER = '%%%%%~=~=~=************=~=~=~%%%%'
+
+CANCELABLE_STATES = [LIVEACTION_STATUS_SCHEDULED, LIVEACTION_STATUS_RUNNING,
+                     LIVEACTION_STATUS_CANCELED]
```
Comment (manasdk) [inline st2common/st2common/constants/action.py]: 'canceled' is a cancelable state?

Comment (lakshmi-kannan) [inline st2common/st2common/constants/action.py]: This is to make canceling a canceled action work with API.

File: st2common/st2common/services/executions.py [hunk 3]
```diff
@@ -41,6 +42,14 @@
 from st2common.models.db.execution import ActionExecutionDB
 from st2common import log as logging
 
+__all__ = [
+    'create_execution_object',
+    'update_execution',
+    'is_execution_canceled',
+    'AscendingSortedDescendantView',
```
Comment (manasdk) [inline st2common/st2common/services/executions.py]: the 2 view classes can be hidden.

File: st2common/st2common/services/action.py [hunk 4]
```diff
@@ -97,3 +100,8 @@ def schedule(liveaction):
     LOG.audit('Action execution scheduled. LiveAction.id=%s, ActionExecution.id=%s' %
               (liveaction.id, execution.id), extra=extra)
     return liveaction, execution
+
+
+def is_action_canceled(liveaction_id):
```
Comment (manasdk) [inline st2common/st2common/services/action.py]: is_execution_canceled?

**Comment by manasdk** [comment]
+1

**Comment by manasdk** [comment]
the testing in this PR is horrible :P

File: st2common/st2common/services/action.py (hunk 4 above)
Comment (lakshmi-kannan) [inline st2common/st2common/services/action.py]: No dude. There is already a method like that in services/execution. See below.

Comment (manasdk) [inline st2common/st2common/services/action.py]: is_action_execution_canceled?

alright whatever works for me.

File: st2common/st2common/constants/action.py (hunk 2 above)
Comment (lakshmi-kannan) [inline st2common/st2common/constants/action.py]: Fixed in c5c16a4

File: st2actions/st2actions/worker.py (hunk 1 above)
Comment (lakshmi-kannan) [inline st2actions/st2actions/worker.py]: Fixed 56c1448. Will add a test case for this.

**Comment by Kami** [comment]
I know I'm a bit late to the game, but I personally think using POST + cancel verb in the URL (e.g. `POST /v1/executions/<id>/CANCEL` is better in this case since we are canceling and not deleting an execution.

**Comment by manasdk** [comment]
multi verb requests make me uneasy. I though the entire point of REST was to reduce verbs :P

**Comment by lakshmi-kannan** [comment]
```POST /v1/executions/<id>/CANCEL``` was my original thought too but it's asymmetric to execution creation. We can support both.

**Pull request merged**